frontend/Rhyolite/Frontend/Form.hs

```diff
@@ -0,0 +1,181 @@
+  return (status, gatedRequest)
+
+withRequestingStatus
+  :: (MonadFix m, MonadHold t m, Requester t m)
+  => Dynamic t (Maybe (Request m a))
+  -> (Dynamic t (RequestStatus (Request m a) (Response m a)) -> m (Event t ()))
+  -> m (Event t (Response m a))
+withRequestingStatus input render = do
+  rec fire <- render status
+      (status, request) <- requestingStatus response fire input
+      response <- requesting request
+  return response
+
+newtype DynValidation t e a = DynValidation { unDynValidation :: Compose (Dynamic t) (Validation e) a }
+
+deriving instance Reflex t => Functor (DynValidation t e)
+deriving instance (Reflex t, Semigroup e) => Applicative (DynValidation t e)
+
+fromDynValidation :: Reflex t => DynValidation t e a -> Dynamic t (Either e a)
+fromDynValidation (DynValidation (Compose v)) = toEither <$> v
+
+toDynValidation :: Reflex t => Dynamic t (Either e a) -> DynValidation t e a
+toDynValidation e = DynValidation $ Compose $ fromEither <$> e
+
+pureDynValidation :: Reflex t => Dynamic t a -> DynValidation t e a
+pureDynValidation a = toDynValidation $ Right <$> a
+
+tagDynValidation :: Reflex t => DynValidation t e b -> Event t a -> Event t b
+tagDynValidation (DynValidation (Compose b)) = push $ \_ -> preview _Success <$> sample (current b)
+
+tagPromptlyDynValidation :: Reflex t => DynValidation t e b -> Event t a -> Event t b
+tagPromptlyDynValidation (DynValidation (Compose b)) = attachPromptlyDynWithMaybe (\b' _ -> preview _Success b') b
+
+manageValidity
+  :: (DomBuilder t m, MonadHold t m, Prerender js m, PerformEvent t m)
+  => Event t () -- When to validate
+  -> (Text -> Either Text a) -- Validation
+  -> m (InputElement EventResult (DomBuilderSpace m) t) -- Render input
+  -> m (InputElement EventResult (DomBuilderSpace m) t, DynValidation t Text a)
+manageValidity validate' validator renderInput = do
+  v@(input, val) <- manageValidation validate' validator renderInput
+  prerender blank $ do
+    let rawEl = _inputElement_raw input
+    performEvent_ $ ffor (tagPromptlyDyn (fromDynValidation val) validate') $ \case
+      Left msg -> do
+        setCustomValidity rawEl msg
+        reportValidity_ rawEl
+      _ -> setCustomValidity rawEl ("" :: Text) -- NOTE setting empty text is how the browser "clears" the error
+  return v
+
+manageValidation
+  :: (DomBuilder t m, MonadHold t m)
+  => Event t () -- When to validate
+  -> (Text -> Either Text a) -- Validation
+  -> m (InputElement EventResult (DomBuilderSpace m) t) -- Render input
+  -> m (InputElement EventResult (DomBuilderSpace m) t, DynValidation t Text a)
+manageValidation validate' validator renderInput = do
+  input <- renderInput
+  let currentVal = current $ value input
+  validatedInput <- fmap (fmap validator) $ buildDynamic (sample currentVal) $ tagPromptlyDyn (value input) validate'
+  return (input, toDynValidation validatedInput)
+
+validateEmail :: Text -> Either Text Text
+validateEmail m = do
```

Comment:
Is there a combinator to set `type="email"` for inputs that use this validator?

Reply:
Similar combinators for min/max/step/password/etc could exist to keep validation lined up with semantic input tags, used by mobile browsers to provide better UIs.